src/index.ts

```diff
@@ -1,7 +1,94 @@
-class ApolloFetcher {
-  public canTest(): boolean {
-      return true;
-  }
+import * as Observable from 'zen-observable';
+
+import { ExecutionResult } from 'graphql';
+
+export interface LinkResult extends ExecutionResult {
+  context?: object;
+  extensions?: any;
 }
 
-export default ApolloFetcher;
+export interface OperationRequest {
+  query: string;
+  context?: any;
+}
+
+export type Linker = (
+  request: OperationRequest,
+  prev?: Link,
+) => Observable<LinkResult | void>;
+
+export default class Link {
+  value: Linker;
+
+  static from(links: Link[]) {
+    return new Link(operation => {
+      return links
+        .reduce(
+          (x, y) => x.concat(y),
+          this instanceof Link ? this : Link.empty(),
+        )
+        .request(operation);
+    });
+  }
+
+  static empty() {
+    return new Link((operation, prev) => {
+      // if starting with empty, just return an Observable
+      if (!prev) return Observable.of(undefined);
+      // otherwise just pass along the prev link
+      return prev.request(operation);
+    });
+  }
+
+  constructor(f) {
+    if (!(this instanceof Link)) return new Link(f);
+    this.value = f;
+  }
+
+  request(
+    operation: OperationRequest,
+    prev?: Link,
+  ): Observable<LinkResult | void> {
+    return this.value(
+      {
+        ...operation,
+        context: operation.context ? operation.context : {},
+      },
+      prev,
+    ).map((result: void | LinkResult) => {
+      if (!result) return;
+      return {
+        ...result,
+        context: result.context ? result.context : operation.context || {},
+      };
+    });
+  }
+
+  map(f: (x: OperationRequest) => OperationRequest) {
+    return new Link(operation => this.request(f(operation)));
+  }
+
+  concat(f: Link) {
```

Comment:
@stubailo I feel like they serve different purposes, but I'd love to talk through it!